File: backend/games/tetris/board.py
"""盤面。1 行を 10bit の整数で持つ（bit x = 列 x）。y=0 が最下段。"""

from __future__ import annotations

from .pieces import BOARD_HEIGHT, BOARD_WIDTH, SHAPES

FULL_ROW = (1 << BOARD_WIDTH) - 1


class Board:
    __slots__ = ("rows",)

    def __init__(self, rows: list[int] | None = None):
        self.rows: list[int] = list(rows) if rows is not None else [0] * BOARD_HEIGHT

    def copy(self) -> "Board":
        return Board(self.rows)

    # --- 参照 ---------------------------------------------------------------
    def is_filled(self, x: int, y: int) -> bool:
        """盤外（左右の壁・床）は埋まっているものとして扱う。"""
        if x < 0 or x >= BOARD_WIDTH or y < 0:
            return True
        if y >= BOARD_HEIGHT:
            return False
        return (self.rows[y] >> x) & 1 == 1

    def collides(self, piece: str, rot: int, x: int, y: int) -> bool:
        shape = SHAPES[piece][rot]
        left = x + shape.min_dx
        if left < 0 or x + shape.max_dx >= BOARD_WIDTH or y + shape.min_dy < 0:
            return True
        rows = self.rows
        for dy, mask in shape.row_masks:
            yy = y + dy
            if yy < BOARD_HEIGHT and rows[yy] & (mask << left):
                return True
        return False

    def is_empty(self) -> bool:
        return not any(self.rows)

    # --- 変更 ---------------------------------------------------------------
    def place(self, piece: str, rot: int, x: int, y: int) -> None:
        shape = SHAPES[piece][rot]
        left = x + shape.min_dx
        for dy, mask in shape.row_masks:
            self.rows[y + dy] |= mask << left

    def clear_lines(self) -> int:
        kept = [r for r in self.rows if r != FULL_ROW]
        cleared = BOARD_HEIGHT - len(kept)
        if cleared:
            self.rows = kept + [0] * cleared
        return cleared

    def add_garbage(self, lines: int, hole: int) -> bool:
        """下から lines 段せり上げる。上から押し出されたら False（top out）。"""
        overflow = any(self.rows[BOARD_HEIGHT - lines:])
        row = FULL_ROW & ~(1 << hole)
        self.rows = [row] * lines + self.rows[: BOARD_HEIGHT - lines]
        return not overflow

    # --- 表示・デバッグ用 -----------------------------------------------------
    def to_strings(self, height: int = 20) -> list[str]:
        """上の行から順に '#'/'.' の文字列。print でそのまま見られる。"""
        return [
            "".join("#" if (self.rows[y] >> x) & 1 else "." for x in range(BOARD_WIDTH))
            for y in range(height - 1, -1, -1)
        ]

    def __repr__(self) -> str:
        return "\n".join(self.to_strings())

File: frontend/src/games/tetris/engine/board.ts
// 盤面。1 行を 10bit の整数で持つ（bit x = 列 x）。y=0 が最下段。backend/tetris/board.py と同じ。

import { BOARD_HEIGHT, BOARD_WIDTH, CELLS, type PieceType, type Rotation } from './pieces'

export const FULL_ROW = (1 << BOARD_WIDTH) - 1

/** 表示用の色。'G' はおじゃま、'' は空き。ルール判定には使わない（Python 版にはない） */
export type CellColor = PieceType | 'G' | ''

export class Board {
  rows: number[]
  /** colors[y][x]。rows と同じように消去・せり上がりで動く */
  colors: CellColor[][]

  constructor(rows?: readonly number[], colors?: readonly (readonly CellColor[])[]) {
    this.rows = rows ? rows.slice() : new Array(BOARD_HEIGHT).fill(0)
    this.colors = colors
      ? colors.map((r) => r.slice())
      : this.rows.map((r) => Array.from({ length: BOARD_WIDTH }, (_, x) => ((r >> x) & 1 ? 'G' : '')))
  }

  copy(): Board {
    return new Board(this.rows, this.colors)
  }

  /** 盤外（左右の壁・床）は埋まっているものとして扱う */
  isFilled(x: number, y: number): boolean {
    if (x < 0 || x >= BOARD_WIDTH || y < 0) return true
    if (y >= BOARD_HEIGHT) return false
    return ((this.rows[y] >> x) & 1) === 1
  }

  collides(piece: PieceType, rot: Rotation, x: number, y: number): boolean {
    for (const [dx, dy] of CELLS[piece][rot]) {
      const cx = x + dx
      const cy = y + dy
      if (cx < 0 || cx >= BOARD_WIDTH || cy < 0) return true
      if (cy < BOARD_HEIGHT && (this.rows[cy] >> cx) & 1) return true
    }
    return false
  }

  isEmpty(): boolean {
    return this.rows.every((r) => r === 0)
  }

  place(piece: PieceType, rot: Rotation, x: number, y: number): void {
    for (const [dx, dy] of CELLS[piece][rot]) {
      this.rows[y + dy] |= 1 << (x + dx)
      this.colors[y + dy][x + dx] = piece
    }
  }

  clearLines(): number {
    const keep = this.rows.map((r) => r !== FULL_ROW)
    const kept = this.rows.filter((_, y) => keep[y])
    const cleared = BOARD_HEIGHT - kept.length
    if (cleared) {
      this.rows = kept.concat(new Array(cleared).fill(0))
      this.colors = this.colors
        .filter((_, y) => keep[y])
        .concat(Array.from({ length: cleared }, () => new Array(BOARD_WIDTH).fill('')))
    }
    return cleared
  }

  /** 下から lines 段せり上げる。上から押し出されたら false（top out） */
  addGarbage(lines: number, hole: number): boolean {
    const overflow = this.rows.slice(BOARD_HEIGHT - lines).some((r) => r !== 0)
    const row = FULL_ROW & ~(1 << hole)
    this.rows = new Array(lines).fill(row).concat(this.rows.slice(0, BOARD_HEIGHT - lines))
    const colorRow = (): CellColor[] => Array.from({ length: BOARD_WIDTH }, (_, x) => (x === hole ? '' : 'G'))
    this.colors = Array.from({ length: lines }, colorRow).concat(this.colors.slice(0, BOARD_HEIGHT - lines))
    return !overflow
  }

  /** デバッグ用: 上の行から '#' / '.' の文字列 */
  toStrings(height = 20): string[] {
    const out: string[] = []
    for (let y = height - 1; y >= 0; y--) {
      let s = ''
      for (let x = 0; x < BOARD_WIDTH; x++) s += (this.rows[y] >> x) & 1 ? '#' : '.'
      out.push(s)
    }
    return out
  }
}
